src/api/logs.rs

```diff
@@ -0,0 +1,318 @@
+                        return Ok(return_400("Commit window must be specified in either seconds `5s` or minutes `1m`"));
+                    }
+
+                    // if the commit window parses to 0 and the value is not 0, 0s or 0m, it's an invalid window
+                    let parsed_window = Config::commit_window_to_seconds(&log.commit_window);
+                    if log.commit_window != "0" 
+                        && log.commit_window != "0s" 
+                        && log.commit_window != "0m" 
+                        && parsed_window == 0 {
+                        return Ok(return_400("Commit window is invalid"));
+                    }
+
+                    let cfg_read = cfg.read().unwrap();
+                    // validate the datastores
+                    for ds_name in &log.datastores {
+                        if cfg_read.datastore.contains_key(ds_name) == false {
+                            return Ok(return_400(&format!("{} is an invalid datastore name", &ds_name)));
+                        }
+                    }
+
+                    // Validate name
+
+                    let mut log_name: String = "".to_string();
+                    if let Some(lg_name) = &log.name {
+                        if lg_name == "" {
+                            return Ok(return_400("Log name cannot be empty."));
+                        }
+                        // validate datastore name uniqueness
+                        if cfg_read.log.contains_key(lg_name) {
+                            return Ok(return_400("Log name already in use"));
+                        }
+                        log_name = lg_name.clone();
+                    }
+
+                    // everything seems ok, create the datastore
+                    let ds_serialized = serde_json::to_string(&log).unwrap();
+
+                    let (tx, rx) = unbounded_channel();
+                    let cfg = Arc::clone(&cfg);
+                    tokio::spawn({
+                        put_object_metabucket(
+                            cfg,
+                            format!("minsql/meta/logs/{}", log_name),
+                            ds_serialized,
+                        )
+                            .map_err(|_| ())
+                            .and_then(move |_| {
+                                log.safe();
+                                let ds_serialized = serde_json::to_string(&log).unwrap();
+
+                                tx.send(ds_serialized).map_err(|_| ())
+                            })
+                            .map(|_| ())
+                            .map_err(|_| ())
+                    });
+
+                    let body_str = rx.map_err(|e| e).map(|x| Chunk::from(x));
+                    let mut response = Response::builder();
+                    response.header(header::CONTENT_TYPE, "application/json");
+
+                    Ok(response.body(Body::wrap_stream(body_str)).unwrap())
```

Comment:
Can't we just wait for the spawned future to complete and then build and return the response, instead of using a channel which seems to be an unnecessary concurrency style here?

Reply:
I really tried to achieve that but was not successful, since we need to read the payload in a future style manner, but return a synchronous streaming body, this was the only solution I was able to do.

Reply:
Here's how you can do it:
https://github.com/dvaldivia/minsql/pull/5

The idea is to make use of Either and split out body parsing into its own function. Writing smaller functions help bring out better abstractions.

Reply:
I only waited to get compilation working - please test the create method!

Reply:
tested, it works